data-donation-sdk/src/commonMain/kotlin/care/data4life/datadonation/donation/publickeyservice/PublicKeyServiceKeyMapper.kt

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2021 D4L data4life gGmbH / All rights reserved.
+ *
+ * D4L owns all legal rights, title and interest in and to the Software Development Kit ("SDK"),
+ * including any intellectual property rights that subsist in the SDK.
+ *
+ * The SDK and its documentation may be accessed and used for viewing/review purposes only.
+ * Any usage of the SDK for other purposes, including usage for the development of
+ * applications/third-party applications shall require the conclusion of a license agreement
+ * between you and D4L.
+ *
+ * If you are interested in licensing the SDK for your own applications/third-party
+ * applications and/or if you’d like to contribute to the development of the SDK, please
+ * contact D4L by email to help@data4life.care.
+ */
+
+package care.data4life.datadonation.donation.publickeyservice
+
+import care.data4life.datadonation.DataDonationSDK
+import care.data4life.datadonation.donation.publickeyservice.model.PublicKeys
+import care.data4life.datadonation.donation.publickeyservice.model.RawKey
+import care.data4life.datadonation.donation.publickeyservice.model.RawKeys
+import care.data4life.datadonation.error.CoreRuntimeError
+
+internal class PublicKeyServiceKeyMapper(
+    private val environment: DataDonationSDK.Environment
+) : PublicKeyServiceContract.Repository.KeyMapper {
+    private fun filterRawKeys(rawKeys: RawKeys): Pair<RawKey, RawKey> {
+        val keys = rawKeys.credentials.filter { rawKey -> environment == rawKey.environment }
+
+        return if (keys.size != 2) {
+            throw CoreRuntimeError.MissingCredentials("Malformed credential source.")
+        } else {
+            Pair(keys.first(), keys.last())
+        }
+    }
+
```

Comment:
This logic of expecting always two keys just to wrap them in a pair/dataClass would break the moment they decide to add another key and i find it fairly unnecessary and limiting. 
We could use a list of public (raw) keys directly on the return method of the Repository with no issues. We can directly filter the array when we need it, no need to limit ourselves on the service beforehand.

Reply:
Nope...actually we are responsible for this particular BackendService so we do not need to assume the worst case here (Thank God!).
I do not get the 2nd part - why you think it is limiting? - The ServiceKeys are always bounded to the environment, which means this mapping has to happen. Clean Architecture following this is a responsibility of the repository (which of course can pull stuff into helper objects), since it is a transformation of an DTO to an Entity.

Reply:
The fact that we are responsible for it doesn't change anything. If we need another credential there, it would break any app using this SDK because it still expects only two, and the only reason why you are forcing it to two is bvecause you want to use Pair. So unnecessary, we can easily keep using lists and remove the "only 2" requirement.

Reply:
We run only in trouble if remove a field or change a field name - adding one is no concern for the serialization. If we need another key for another service we need to touch the SDK anyways, so I do not see the point here. This also goes for renaming or removing stuff.
Also I do not need to use Pair here...but in this case it is convenient. I could use also a list...but it does not change the fact, that there are 2 keys.

Reply:
this method throws an error if there are 3 keys, meaning that the sdk stops working. we cant update sdk in apps in production so "we need to touch the SDK anyways" is not a valid point, but it will make the not updated apps not work. Please change to a more flexible list.

Reply:
You are pointing on something, but it is not related to here. Alter the fixture on #26 and run the integration tests to see for yourself. There is a change needed at the `KeyDomainSerializer`, to ignore unknown key fields...totally forgot about that, since it has been a while since I touched this PR.

Reply:
👍